### PR TITLE
Gh 89 metrics

### PR DESCRIPTION
**Pull request by go159lab**
- Add springboot actuator metrics and promethous.
- Add @Timed to capture the data from Micrometer's @Timed annotation into the Prometheus metrics store/registry
- Add MeterRegistry Counters to report a single metric, a count. Which allows you to increment by a fixed amount, which must be positive.

### CODE REVIEW

File: gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    template:
+      metadata:
+        annotations:
+          openshift.io/generated-by: OpenShiftNewApp
+        creationTimestamp: null
+        labels:
+          deployment: prometheus
+      spec:
+        containers:
+        - image: prom/prometheus@sha256:3e6f1ebc047ec30971c92e86aa3b879b07d39327a7744cce8aa0b1076d4c94d6
```
Comment (r32575) [inline gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:63]: Tying it to a particular image. Don't know if you want to or not?

**Review by r32575 — CHANGES_REQUESTED** [review]

File: gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml [hunk 2]
```diff
@@ -0,0 +1,121 @@
+kind: Template
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:1]: The Prometheus setup is currently dependent on gaas-rest as it collects metrics from the /actuator/prometheus path of the API.

File: gaffer-as-a-service/gaas-rest/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+version: "2"
+services:
+  gaas-rest:
+    image: "kitkat520/api"
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/docker-compose.yml:4]: done

File: gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+  metadata:
+    annotations:
+      openshift.io/host.generated: "true"
+    labels:
+      app: prometheus
+      app.kubernetes.io/component: prometheus
+      app.kubernetes.io/instance: prometheus
+    name: prometheus
+    resourceVersion: "29460365"
+    uid: 12ab184b-ad9f-4b5a-bf53-240e37dcad55
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:96]: done

File: gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml (hunk 1 above)
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:63]: done

File: gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (r32575) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:1]: `actuator/prometheus` sounds like a generic API address that you could for example use for the `gaffer-controller` as well.

[the remark this answers is not in the document]
Comment (r32575) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:1]: I'd prefer it if you created the module as part of this ticket

[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml:1]: This work is being done under ticket 89

[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml]: Prometheus has been refactored into geas-common and README updated.

There are some additional metrics which need to be added before merging which are being worked on under #89

[the remark this answers is not in the document]
Comment (r32575) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml]: Sorry, the comment was hidden and I didn't see it until now. I wanted Prometheus to be in a different module that sits alongside `gaffer-as-a-service` as it isn't tied to `gaas-rest` or `gaffer-controller`. For example `gaffer-as-a-service-tools`

**Comment by go159lab** [comment]
Hi @r32575, @t92549 
Gh 89 metrics is ready to be reviewed.

**Comment by t92549** [comment]
Happy to merge when this is solved https://github.com/gchq/gaffer-experimental/pull/90#discussion_r644150089

File: gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/deploy/openshift-prometheus.yaml]: Prometheus has been moved to `gaffer-as-a-service-tools` we have also tested it with geas-rest in openShift and it's working.

**Comment by rocky341** [comment]
> Happy to merge when this is solved [#90 (comment)](https://github.com/gchq/gaffer-experimental/pull/90#discussion_r644150089)

@t92549 Updated so should now be resolved

**Review by t92549 — APPROVED** [review]

**Review by r32575 — APPROVED** [review]

**Pull request merged**